feat(timeselect): Add inline dp-time-select

isaacplmann opened the pull request:
I've added an inline `<dp-time-select>` component.  Here's the config object:

```ts
export interface ITimeSelectConfig extends ICalendar {
  hours12Format?: string; // 'hh'
  hours24Format?: string; // 'HH'
  meridiemFormat?: string; // 'A'
  minutesFormat?: string; // 'mm'
  minutesInterval?: number; // 1
  secondsFormat?: string; // 'ss'
  secondsInterval?: number; // 1
  showSeconds?: boolean; // false
  showTwentyFourHours?: boolean; // false
  timeSeparator?: string; // ':'
}
```

Next step is to add values to the ECalendarType enum: `Time` and `DayTime` (or `DateTime`) and then add the appropriate integrations in the `<dp-day-picker>`.  I'll work on `Time` first, since it is the simpler one.

Discussion points:
1. I can foresee some ambiguity with the `min` and `max` values in the config object.  For a `Time` calendar type, the `min` and `max` should probably have the day portion of the datetime ignored and only compare the time portion with the value in the `<dp-time-select>`.  However, for a `DayTime` calendar type, the day portion needs to be included in the calculation.  I think it is clear enough from the context what `min` and `max` should mean, but, if you want, I can add `minTime` and `maxTime` properties to the `ITimeSelectConfig`.

2. It would probably be useful to update the demo code to hide irrelevant config options based on the currently selected mode.  I was thinking of creating some kind of look up table to hide/show config options as the mode changes.  I'd be happy to work on that if you think it's a good idea.

isaacplmann commented:
Starts to implement #5

vlio20 commented:
Hi @isaacplmann,
1. I think it is a good idea to support `minTime/maxTime` if user don't want to allow specific hours for a day.
2. sure - sounds good.

Few questions and points that have to be added: 
1. Disabled options.
2. there should be an option to have validation errors for min and max time (not only disabled state).
3. Material theme also should be supported.
4. Can it be placed next to day picker? - I think it should (how will it look?).
5. what about a directive (sees like many of the users a using this option (according to open issues), will you also add it?

isaacplmann commented:
3. What else should I do for the material theme?  I change the buttons to being circles that appear on hover.
4. Yes, I'm planning on implementing that.
5. Yes, I'll add a directive version.

isaacplmann commented:
Todo list:

- [ ] `minTime`/`maxTime`
- [ ] show only relevant config options in demo
- [ ] disabled options
- [ ] validation errors for `minTime`/`maxTime`
- [ ] `Time` version of `dp-day-picker` (opens a `time-select`)
- [ ] inline `DayTime` - (`time-select` and `day-calendar` side by side)
- [ ] `DayTime` version of `dp-day-picker`
- [ ] directive version of `Time` calendar type
- [ ] directive version of `DayTime` calendar type

vlio20 commented:
3. Can you please share a screenshot?

isaacplmann commented:
Material theme (hovering over the bottom middle button)
![screen shot 2017-06-26 at 9 20 08 am](https://user-images.githubusercontent.com/861504/27541110-bda8b6ea-5a50-11e7-98fa-152477f1d07d.png)

No theme
![screen shot 2017-06-26 at 9 20 20 am](https://user-images.githubusercontent.com/861504/27541134-cd682bec-5a50-11e7-8898-7aaf88d7e04c.png)

isaacplmann commented:
Travis CI is failing, but I think that's happening in code that I haven't touched.  Any thoughts?  When I run `npm test` all the tests pass.

vlio20 commented:
will check

vlio20 commented:
@isaacplmann, can you also deploy the demo to your forked project?

vlio20 commented on src/app/common/services/utils/utils.service.ts:
```diff
@@ -6,6 +6,14 @@ import {Moment, unitOfTime} from 'moment';
 import {CalendarValue} from '../../types/calendar-value';
 import {IDate} from '../../models/date.model';
 
+export type DateValidatorFn = (inputVal: CalendarValue) => {[key: string]: any};
```
why any?

vlio20 commented on src/app/demo/demo/demo.component.html:
```diff
@@ -127,6 +128,38 @@
+            [theme]="material ? 'dp-material' : ''"
+            [(ngModel)]="date"
+            (ngModelChange)="log($event)"
+            [minDate]="validationMinDate"
```
why the time component has min/maxDate validations?

isaacplmann replied:
You're right, I should put the `minDate` and `maxDate` validation only on the `DayTime` component and not on the `Time` component.

vlio20 commented on src/app/time-select/time-select.component.html:
```diff
@@ -0,0 +1,28 @@
+<div class="dp-time-select-container">
+  <ul class="controls">
+    <li class="control">
+      <button [disabled]="!shouldShowIncrease('hour')" class="control-up" (click)="increase('hour')"></button>
```
all classes should be prefixed with dp

vlio20 commented on src/app/time-select/time-select.component.html:
```diff
@@ -0,0 +1,28 @@
+  <ul class="controls">
+    <li class="control">
+      <button [disabled]="!shouldShowIncrease('hour')" class="control-up" (click)="increase('hour')"></button>
+      <span class="control--hours">{{hours}}</span>
```
I am not using BEM, so please remove

vlio20 commented on src/app/time-select/time-select.component.less:
```diff
@@ -0,0 +1,55 @@
+@import '../common/styles/variables';
+
+& {
+  .controls {
```
please add line break after each class

vlio20 commented on src/app/time-select/time-select.component.less:
```diff
@@ -0,0 +1,55 @@
+    font-size: 18px;
+    font-weight: bold;
+    cursor: pointer;
+    &:before {
```
: =-> ::

vlio20 commented on src/app/time-select/time-select.component.less:
```diff
@@ -0,0 +1,55 @@
+    }
+  }
+  .control-up:before {
+    content: '︿';
```
We shouldn't use this type of characters as they are probably not supported on linux or other OSs. Please use pure CSS.

vlio20 commented on src/app/time-select/time-select.component.less:
```diff
@@ -0,0 +1,55 @@
+  }
+  .control-up:before {
+    content: '︿';
+    bottom: 0.2em;
```
why em? please use px

vlio20 commented on src/app/time-select/time-select.component.ts:
```diff
@@ -0,0 +1,189 @@
+      const {minDate, maxDate, minTime, maxTime} = changes;
+      this.init();
+
+      if (minDate || maxDate || minTime || maxTime) {
```
please push this check into the `initValidatiors` method

isaacplmann replied:
That check doesn't make sense in the `initValidators` method.  It's checking to see if those properties have been changed - which only makes sense in `ngOnChanges`. See https://github.com/vlio20/angular-datepicker/blob/master/src/app/day-calendar/day-calendar.component.ts#L119

vlio20 replied:
you are totally right - my mistake

vlio20 commented on src/app/time-select/time-select.component.ts:
```diff
@@ -0,0 +1,189 @@
+
+  @Input() config: ITimeSelectConfig;
+  @Input() displayDate: SingleCalendarValue;
+  @Input() minDate: Moment;
```
we should also support strings, that means that this component should also receive a format as an input.

vlio20 commented on src/app/time-select/time-select.service.ts:
```diff
@@ -0,0 +1,128 @@
+    secondsInterval: 1,
+    showSeconds: false,
+    showTwentyFourHours: false,
+    timeSeparator: ':',
```
redundant ,

isaacplmann replied:
I'm modeling the `DEFAULT_CONFIG` style off of `day-calendar.service` and `month-calendar.service`.  What's redundant?

isaacplmann replied:
Ah, you're referring to the comma.  Ok, I'll remove it.

vlio20 replied:
yup

vlio20 commented on src/app/time-select/time-select.service.ts:
```diff
@@ -0,0 +1,128 @@
+      maxTime: this.utilsService.onlyTime(config && config.maxTime),
+      minTime: this.utilsService.onlyTime(config && config.minTime),
+    };
+    return {...this.DEFAULT_CONFIG, ...this.utilsService.clearUndefined(config), ...timeConfigs};
```
please break to lines

vlio20 commented on src/app/time-select/time-select.service.ts:
```diff
@@ -0,0 +1,128 @@
+  getConfig(config: ITimeSelectConfig): ITimeSelectConfig {
+    const timeConfigs = {
+      maxTime: this.utilsService.onlyTime(config && config.maxTime),
+      minTime: this.utilsService.onlyTime(config && config.minTime),
```
redundant ,
I will add lint rule for that in the near future

isaacplmann commented:
How do you deploy the demo?  I see a script to build the demo, but not to deploy it.

vlio20 commented:
1. go to release branch (dist folder is not ignored there) 
2. `run npm run build:demo`
3. run: `git subtree push --prefix dist origin gh-pages`

vlio20 commented:
@isaacplmann, I pushed a fix to master for the travis build there is still an issue with one e2e tests which is not related to your code.
I will try to fix it later.

isaacplmann commented:
Demo is up: https://isaacplmann.github.io/ng2-date-picker/

isaacplmann commented:
Made the buttons css only - and I think they look better too.

isaacplmann commented:
Created `daytimeInline` and `daytimePicker` modes.  Next I'll work on `timeDirective` and `daytimeDirective`.

isaacplmann commented:
Directive versions are finished.  I've updated the demo too.

isaacplmann commented:
You can check out the new docs here: https://github.com/isaacplmann/ng2-date-picker/tree/timepicker

I'm done with my changes - I'm just waiting on your review.

vlio20 commented:
can you please add the demo again - seems that it is not working any more ...

isaacplmann commented:
Here's a hosted version of the demo: https://isaacplmann.github.io/ng2-date-picker/
I also just ran `npm start` on my local copy and it served up the demo.  What isn't working for you?

vlio20 commented on src/app/time-select/time-select.service.ts:
```diff
@@ -0,0 +1,128 @@
+import {Injectable} from '@angular/core';
+import {Moment} from 'moment';
+import * as moment from 'moment';
+
```
redundant

vlio20 commented:
@isaacplmann, the demo looks great!!! 
one thing though is with the no theme appearance - the button icon is not in the middle: 
![Imgur](http://i.imgur.com/FJNcVs2.png)

Also I would prefer to make it more slim: 
![Imgur](http://i.imgur.com/5o6SbDP.png)

vlio20 commented:
another question: what is `package-lock.json`?  I was using shrinkwrap.

isaacplmann commented:
`package-lock.json` is a file generated by `npm@5.x`.  It's the next version of shrinkwrap.  I'll remove it from the PR, I must have accidentally included it.

isaacplmann commented:
I'll fix the e2e tests and add some for the new components tomorrow.

isaacplmann commented:
Fixed e2e tests and added some for `timePicker` and `daytimePicker`.

vlio20 commented:
This is awesome! Can you just make sure that the default state in the demo is the date picker with the time selector?

isaacplmann commented:
You want the default `pickerMode` to be `daytimePicker`?

isaacplmann commented:
If that's the case, then the default Date Format should be set to `DD-MM-YYYY HH:mm:ss`

vlio20 commented:
Yes, it's shows better the capabilities of this component. 
There probably should be default format for each type of calendar - so for there date time picker it should be `Yes `DD-MM-YYYY HH:mm:ss` as you mentioned.

isaacplmann commented:
I updated `pickerMode` to use `daytimePicker` by default and fixed the e2e tests that were broken by that change.  I'll leave modifying the `config.format` when the `pickerMode` changes to you.

vlio20 commented:
cool, so you are done?

isaacplmann commented:
Yep.  All done.

vlio20 commented:
I will release a version soon. thanks again.

isaacplmann commented:
Thank you!

> On Jun 29, 2017, at 12:09 PM, Vlad Ioffe <notifications@github.com> wrote:
> 
> I will release a version soon. thanks again.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/vlio20/angular-datepicker/pull/109#issuecomment-312015099>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AA0lQI2UwgNc8ohn0HcfM-sH3gUqC1fSks5sI8w2gaJpZM4OD-YC>.
>
